File: src/spotsolve/patches.py
"""Decompose a set of emitters into small, jointly-fittable patches.

Emitters whose ~sigma-scale supports touch are grouped into one patch and
fit jointly. Emitters just outside a patch (within a further halo radius)
are held fixed and folded into the model as a constant contribution, so
flux is not double-counted at patch borders and the joint Hessian stays
small (bounded by K_max).

Choosing halo_radius_factor
---------------------------
An emitter that is neither free in a patch nor in its frozen halo is, from
that patch's point of view, not in the model at all -- and the patch's free
background `b` is the only parameter that can absorb it. So the halo radius
has to be large enough that what it excludes is genuinely negligible against
the BACKGROUND, not merely small against a bead peak.

Measured on random fields with bead-matched brightness (peaks 94-198 e-,
background 4 e-), worst-case leak into a patch from emitters outside its
halo:

    halo_radius_factor   max leak   p99 leak   frozen emitters per patch
            3.0            3.10       2.40              2.8
            4.0            0.155      0.086             3.7
            5.0            0.001      0.001             4.6

At the old 3.0 a patch could be handed an unmodelled pedestal of 3.1 e- on
a 4 e- background. That is not a small perturbation: with the halo removed
entirely the same patch fit drives `b` to 57 e- and its precision and
recall fall from 1.00/0.75 to 0.62/0.62. 5.0 buys the margin for about two
extra frozen emitters per patch, and frozen emitters cost only one rendered
constant each -- they do not enter the Hessian. It also keeps the halo at
least as wide as the 4-sigma truncation `detect.render_model` uses, so the
halo never omits flux the global model includes.
"""

import numpy as np
from scipy.sparse.csgraph import connected_components
from scipy.spatial import cKDTree

from . import psf
from .structs import Patch

__all__ = ["build_patches", "build_halo_image", "patch_grids"]


def _bbox_for(ys, xs, pad, img_shape):
    y0 = max(0, int(np.floor(ys.min() - pad)))
    x0 = max(0, int(np.floor(xs.min() - pad)))
    y1 = min(img_shape[0], int(np.ceil(ys.max() + pad)) + 1)
    x1 = min(img_shape[1], int(np.ceil(xs.max() + pad)) + 1)
    return y0, x0, y1, x1


def build_patches(
    positions,
    sigma,
    img_shape,
    link_radius_factor=2.5,
    halo_radius_factor=5.0,
    bbox_pad_factor=3.0,
    k_max=12,
):
    """positions: (N,2) array of (y,x). Returns a list of Patch."""
    positions = np.asarray(positions, dtype=float)
    n = positions.shape[0]
    if n == 0:
        return []

    tree = cKDTree(positions)
    link_r = link_radius_factor * sigma
    pairs = tree.query_pairs(r=link_r, output_type="ndarray")

    if pairs.shape[0] == 0:
        adj = np.zeros((n, n), dtype=bool)
    else:
        adj = np.zeros((n, n), dtype=bool)
        adj[pairs[:, 0], pairs[:, 1]] = True
        adj[pairs[:, 1], pairs[:, 0]] = True

    n_comp, labels = connected_components(adj, directed=False)

    patches = []
    halo_r = halo_radius_factor * sigma
    bbox_pad = bbox_pad_factor * sigma

    for c in range(n_comp):
        idx = np.nonzero(labels == c)[0]
        if idx.shape[0] > k_max:
            for sub in _split_component(idx, positions, k_max):
                patches.append(
                    _finalize_patch(sub, positions, tree, halo_r, bbox_pad, img_shape)
                )
        else:
            patches.append(
                _finalize_patch(idx, positions, tree, halo_r, bbox_pad, img_shape)
            )
    return patches


def _finalize_patch(idx, positions, tree, halo_r, bbox_pad, img_shape):
    ys, xs = positions[idx, 0], positions[idx, 1]
    y0, x0, y1, x1 = _bbox_for(ys, xs, bbox_pad, img_shape)

    # frozen halo: emitters not in idx, within halo_r of the bbox
    cy, cx = (y0 + y1) / 2.0, (x0 + x1) / 2.0
    diag = np.hypot(y1 - y0, x1 - x0) / 2.0
    cand = tree.query_ball_point([cy, cx], diag + halo_r)
    cand = np.array([i for i in cand if i not in set(idx.tolist())], dtype=int)
    if cand.size:
        # keep only those actually within halo_r of the bbox rectangle
        py = np.clip(positions[cand, 0], y0, y1 - 1)
        px = np.clip(positions[cand, 1], x0, x1 - 1)
        d = np.hypot(positions[cand, 0] - py, positions[cand, 1] - px)
        cand = cand[d <= halo_r]

    return Patch(indices=idx, frozen_indices=cand, y0=y0, x0=x0, y1=y1, x1=x1)


def _split_component(idx, positions, k_max):
    """Recursively bisect an oversized component by a spatial median split
    (along its longer axis) until every piece is <= k_max. This is a proxy
    for "cut the weakest graph edge": splitting along the axis of greatest
    spread tends to cut through the sparsest part of a spatially clustered
    component."""
    if idx.shape[0] <= k_max:
        return [idx]
    pts = positions[idx]
    spread = pts.max(axis=0) - pts.min(axis=0)
    axis = int(np.argmax(spread))
    order = idx[np.argsort(pts[:, axis])]
    mid = order.shape[0] // 2
    left, right = order[:mid], order[mid:]
    return _split_component(left, positions, k_max) + _split_component(
        right, positions, k_max
    )


def build_halo_image(positions, amplitudes, frozen_indices, sigma, yy, xx, y0, x0):
    """Render the constant (parameter-free) contribution from frozen
    (out-of-patch) emitters. The patch's own background `b` is a free
    parameter handled separately in the patch's theta vector.

    `yy, xx` are the patch's LOCAL grids (see patch_grids); `positions` are
    global, so they are converted to local coordinates here via (y0, x0)
    -- the same origin passed to patch_grids for this patch.
    """
    if frozen_indices.size == 0:
        return 0.0
    A = np.asarray(amplitudes)[frozen_indices]
    pos = np.asarray(positions)[frozen_indices]
    theta = psf.pack(0.0, A, pos[:, 0] - y0, pos[:, 1] - x0)
    return psf.model(theta, yy, xx, sigma)


def patch_grids(patch):
    """Local (0-based) pixel-center coordinate grids for this patch, i.e.
    grid[0,0] corresponds to global pixel (patch.y0, patch.x0) -- matching
    both `data_img[patch.y0:patch.y1, patch.x0:patch.x1]` (sub-image
    extraction) and how theta positions are built/stored (local coords,
    with patch.y0/x0 added back only when writing into the global
    positions array). Do not switch this to global coordinates without
    also removing every local<->global offset elsewhere."""
    h, w = patch.shape
    yy, xx = np.mgrid[0:h, 0:w]
    return yy * 1.0, xx * 1.0

File: rust/spotsolve-core/src/patches.rs
//! Group emitters into bounded joint-fit patches with fixed neighboring light.
//!
//! `LINK_FACTOR` sets connectivity, `K_MAX` caps free emitters, `HALO_FACTOR`
//! includes frozen neighbors, and `BBOX_PAD` supplies pixel context. Frozen
//! emitters contribute to the model but not to the fitted covariance.
//! These radii use the reference sigma; wide sources can extend beyond them.
//! Geometry fixtures verify the partition and coordinate conventions.

use crate::grid::EmitterGrid;

pub const LINK_FACTOR: f64 = 2.5;
pub const HALO_FACTOR: f64 = 5.0;
pub const BBOX_PAD: f64 = 3.0;
pub const K_MAX: usize = 12;

/// A pixel bounding box: `y0`/`x0` inclusive, `y1`/`x1` exclusive.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct BBox {
    pub y0: usize,
    pub x0: usize,
    pub y1: usize,
    pub x1: usize,
}

impl BBox {
    #[inline]
    pub fn h(&self) -> usize {
        self.y1 - self.y0
    }
    #[inline]
    pub fn w(&self) -> usize {
        self.x1 - self.x0
    }
    #[inline]
    pub fn n_pixels(&self) -> usize {
        self.h() * self.w()
    }
}

/// A group of emitters fitted jointly, plus its pixel bounding box.
///
/// `indices` are the FREE emitters; `frozen` are neighbours just outside the
/// group, held fixed and folded into the model as a constant halo so flux is
/// not double-counted at patch borders and the joint Hessian stays bounded.
#[derive(Clone, Debug)]
pub struct Patch {
    pub indices: Vec<u32>,
    pub frozen: Vec<u32>,
    pub bbox: BBox,
}

fn bbox_for(ys: &[f64], xs: &[f64], pad: f64, h: usize, w: usize) -> BBox {
    let (mut ymin, mut ymax) = (f64::INFINITY, f64::NEG_INFINITY);
    let (mut xmin, mut xmax) = (f64::INFINITY, f64::NEG_INFINITY);
    for i in 0..ys.len() {
        ymin = ymin.min(ys[i]);
        ymax = ymax.max(ys[i]);
        xmin = xmin.min(xs[i]);
        xmax = xmax.max(xs[i]);
    }
    BBox {
        y0: (ymin - pad).floor().max(0.0) as usize,
        x0: (xmin - pad).floor().max(0.0) as usize,
        y1: (((ymax + pad).ceil() as i64 + 1).max(0) as usize).min(h),
        x1: (((xmax + pad).ceil() as i64 + 1).max(0) as usize).min(w),
    }
}

struct UnionFind(Vec<u32>);

impl UnionFind {
    fn new(n: usize) -> Self {
        UnionFind((0..n as u32).collect())
    }
    fn find(&mut self, mut i: u32) -> u32 {
        while self.0[i as usize] != i {
            self.0[i as usize] = self.0[self.0[i as usize] as usize];
            i = self.0[i as usize];
        }
        i
    }
    fn union(&mut self, a: u32, b: u32) {
        let (ra, rb) = (self.find(a), self.find(b));
        if ra != rb {
            // Always attach the larger root to the smaller, so the
            // representative of a component is its lowest member regardless of
            // the order pairs arrive in.
            let (lo, hi) = if ra < rb { (ra, rb) } else { (rb, ra) };
            self.0[hi as usize] = lo;
        }
    }
}

/// Group `n` emitters into jointly-fittable patches.
///
/// Components larger than `k_max` are bisected; see [`split_component`].
pub fn build_patches(
    pos: &[f64],
    n: usize,
    sigma: f64,
    h: usize,
    w: usize,
    k_max: usize,
) -> Vec<Patch> {
    if n == 0 {
        return Vec::new();
    }
    let halo_r = HALO_FACTOR * sigma;
    let link_r = LINK_FACTOR * sigma;
    let grid = EmitterGrid::build(pos, n, h, w, halo_r);

    let mut uf = UnionFind::new(n);
    let mut cand = Vec::new();
    for i in 0..n {
        grid.query_disc(pos, pos[2 * i], pos[2 * i + 1], link_r, &mut cand);
        for &j in &cand {
            if j as usize != i {
                uf.union(i as u32, j);
            }
        }
    }

    // Components, each with its members in ascending index order -- which is
    // what `np.nonzero(labels == c)` produces, and what `split_component`
    // assumes.
    let mut root_of: Vec<u32> = (0..n).map(|i| uf.find(i as u32)).collect();
    let mut order: Vec<u32> = (0..n as u32).collect();
    order.sort_by_key(|&i| (root_of[i as usize], i));
    let mut patches = Vec::new();
    let mut start = 0usize;
    while start < order.len() {
        let r = root_of[order[start] as usize];
        let mut end = start;
        while end < order.len() && root_of[order[end] as usize] == r {
            end += 1;
        }
        let comp: Vec<u32> = order[start..end].to_vec();
        if comp.len() > k_max {
            for sub in split_component(&comp, pos, k_max) {
                patches.push(finalize(sub, pos, &grid, halo_r, BBOX_PAD * sigma, h, w));
            }
        } else {
            patches.push(finalize(comp, pos, &grid, halo_r, BBOX_PAD * sigma, h, w));
        }
        start = end;
    }
    root_of.clear();
    patches
}

/// Recursively bisect an oversized component by a spatial median split along
/// its longer axis, until every piece is at most `k_max`.
///
/// A proxy for "cut the weakest graph edge": splitting along the axis of
/// greatest spread tends to cut through the sparsest part of a spatially
/// clustered component.
fn split_component(idx: &[u32], pos: &[f64], k_max: usize) -> Vec<Vec<u32>> {
    if idx.len() <= k_max {
        return vec![idx.to_vec()];
    }
    let (mut ylo, mut yhi) = (f64::INFINITY, f64::NEG_INFINITY);
    let (mut xlo, mut xhi) = (f64::INFINITY, f64::NEG_INFINITY);
    for &i in idx {
        ylo = ylo.min(pos[2 * i as usize]);
        yhi = yhi.max(pos[2 * i as usize]);
        xlo = xlo.min(pos[2 * i as usize + 1]);
        xhi = xhi.max(pos[2 * i as usize + 1]);
    }
    let axis = if (yhi - ylo) >= (xhi - xlo) { 0 } else { 1 };
    let mut order = idx.to_vec();
    // Stable, with the emitter index as an explicit tiebreak. numpy's argsort
    // defaults to an unstable quicksort, so matching it on ties is neither
    // possible nor desirable; being deterministic is.
    order.sort_by(|&a, &b| {
        let (va, vb) = (pos[2 * a as usize + axis], pos[2 * b as usize + axis]);
        va.partial_cmp(&vb)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then(a.cmp(&b))
    });
    let mid = order.len() / 2;
    let mut out = split_component(&order[..mid], pos, k_max);
    out.extend(split_component(&order[mid..], pos, k_max));
    out
}

fn finalize(
    idx: Vec<u32>,
    pos: &[f64],
    grid: &EmitterGrid,
    halo_r: f64,
    bbox_pad: f64,
    h: usize,
    w: usize,
) -> Patch {
    let ys: Vec<f64> = idx.iter().map(|&i| pos[2 * i as usize]).collect();
    let xs: Vec<f64> = idx.iter().map(|&i| pos[2 * i as usize + 1]).collect();
    let bbox = bbox_for(&ys, &xs, bbox_pad, h, w);
    let mut cand = Vec::new();
    let frozen = frozen_around(pos, grid, &bbox, halo_r, &idx, &mut cand);
    Patch {
        indices: idx,
        frozen,
        bbox,
    }
}

/// Emitters within `halo_r` of the bbox *rectangle* (not of its centre), less
/// those already free in `exclude`. Ascending index order.
fn frozen_around(
    pos: &[f64],
    grid: &EmitterGrid,
    bbox: &BBox,
    halo_r: f64,
    exclude: &[u32],
    cand: &mut Vec<u32>,
) -> Vec<u32> {
    grid.query_rect(
        bbox.y0 as f64,
        bbox.x0 as f64,
        (bbox.y1 - 1) as f64,
        (bbox.x1 - 1) as f64,
        halo_r,
        cand,
    );
    cand.iter()
        .copied()
        .filter(|i| !exclude.contains(i))
        .filter(|&i| dist_to_bbox(pos, i as usize, bbox) <= halo_r)
        .collect()
}

/// Distance from an emitter to the closed pixel rectangle `[y0, y1-1] x [x0, x1-1]`.
fn dist_to_bbox(pos: &[f64], i: usize, b: &BBox) -> f64 {
    let (y, x) = (pos[2 * i], pos[2 * i + 1]);
    let py = y.clamp(b.y0 as f64, (b.y1 - 1) as f64);
    let px = x.clamp(b.x0 as f64, (b.x1 - 1) as f64);
    ((y - py).powi(2) + (x - px).powi(2)).sqrt()
}
